### PR TITLE
Remove unnecessary restcall which caused problems

### PR DESCRIPTION
**Pull request by F4lka**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
~- [x] Server: I added multiple integration tests (Spring) related to the features~
~- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)~
~- [x] Server: I implemented the changes with a good performance and prevented too many database calls~
- [x] Server: I documented the Java code using JavaDoc style.
~- [x] Client: I added multiple integration tests (Jest) related to the features~
~- [x] Client: I added `authorities` to all new routes and check the course groups for displaying~ navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
~- [x] Client: I added multiple screenshots/screencasts of my UI changes~
~- [x] Client: I translated all the newly inserted strings into German and English~

### Steps for Testing

Focus on the student accounts for testing, but please check also the others(tutor, instructor, admin).

Testing on test server:
1. Log into Artemis
2. Navigate to Course Management
3. Select a course
4. Create a programming exercise
5. Navigate to course overview
6. Select the course 
7. Open the details page of the programming exercise

-> No error should occur here, check the network tab of the browser developer tools

Local testing:
1. Run artemis in your IDE
2. Login into Artemis
3. Navigate to Course Management
4. Select a course
5. Create a programming exercise without a connection to the vcs and ci 
-> normal creation procedure as for a normal programming exercise, but before you click on generate tick the checkbox "Setup without connection to the VCS and CI"

Screenshot from 2020-04-30 18-23-16

6. Navigate to course overview
7. Select the course
8. Open the details page of the programming exercise

-> No error should occur here, check the network tab of the browser developer tools
-> Two buttons (submission and result) should be visible
Screenshot from 2020-04-30 18-27-55

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java [hunk 1]
```diff
@@ -372,4 +372,15 @@ private void setAssignedTeamIdForExerciseAndUser(Exercise exercise, User user) {
...
     }
+
+    /**
+     * Sets the transient attribute "testRepositoryUrl" if the exercises is a programming exercise
+     *
+     * @param exercise the exercise for which to set the url
+     */
+    public void setTestRepositoryUrlForProgrammingExercise(Exercise exercise) {
+        if (exercise instanceof ProgrammingExercise) {
+            exercise.setTestRepositoryUrlTransient(programmingExerciseService.findByIdWithEagerStudentParticipations(exercise.getId()).getTestRepositoryUrl());
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java:383]: Why do you fetch again? I think you can simply cast the java object here

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java [hunk 2]
```diff
@@ -172,6 +172,9 @@
     @Transient
     private Long studentAssignedTeamIdTransient; // id of the team that the logged-in user is assigned to (only relevant if team mode is enabled)
 
+    @Transient
+    private String testRepositoryUrlTransient;
+
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:177]: You misunderstood me. Do not set the string again here. Use a transient Boolean in ProgrammingExercise called isLocalSimulation (or similar) and only set this to true if the url includes artemislocal (Same Check as on the client)

File: src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java (hunk 1 above)
Comment (F4lka) [reply src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java]: Yes, you are right, thank you for the tip

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java (hunk 2 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:177]: please still move this to `ProgrammingExercise.java`

File: src/main/webapp/app/entities/exercise.model.ts [hunk 3]
```diff
@@ -88,6 +88,7 @@ export abstract class Exercise implements BaseEntity {
     public numberOfMoreFeedbackRequests?: number;
     public numberOfOpenMoreFeedbackRequests?: number;
     public studentAssignedTeamId?: number;
+    public isLocalSimulation?: boolean;
```
Comment (krusche) [inline src/main/webapp/app/entities/exercise.model.ts:91]: move to ProgrammingExercise

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java (hunk 2 above)
Comment (F4lka) [reply src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:177]: Alright, i updated the transient property

File: src/main/webapp/app/entities/exercise.model.ts (hunk 3 above)
Comment (F4lka) [reply src/main/webapp/app/entities/exercise.model.ts]: i moved it to the programming exercise

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java [hunk 4]
```diff
@@ -777,4 +777,16 @@ public long countSubmissionsByCourseIdSubmitted(Long courseId) {
         log.debug("countSubmissionsByCourseIdSubmitted took " + (System.currentTimeMillis() - start) + "ms");
         return count;
     }
+
+    /**
+     * Sets the transient attribute "testRepositoryUrl" if the exercises is a programming exercise
+     *
+     * @param exercise the exercise for which to set the url
+     */
+    public void setTestRepositoryUrlForProgrammingExercise(Exercise exercise) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java:786]: please rename the method, it does not reflect what is done here.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java [hunk 5]
```diff
@@ -777,4 +777,16 @@ public long countSubmissionsByCourseIdSubmitted(Long courseId) {
...
+        if (exercise instanceof ProgrammingExercise) {
+            boolean isLocalSimulation = (((ProgrammingExercise) exercise).getTestRepositoryUrl()).contains("artemislocalhost");
+            ((ProgrammingExercise) exercise).setIsLocalSimulation(isLocalSimulation);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java:789]: I would only set this value if it is a local simulation

**Comment by kloessst** [comment]
Successfully tested the changes on test server 3. No unnecessary request was send to the server when the exercise details page was visited as admin, instructor, tutor or student.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java (hunk 4 above)
Comment (F4lka) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java]: i renamed the method to checksAndSetsIfProgrammingExerciseIsLocalSimulation

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java (hunk 5 above)
Comment (F4lka) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java]: now the value will be only set if it is a local simulation

**Pull request merged**